### ISSUE TITLE
Lesion tracking using DL methods

### ISSUE THREAD

**Issue by plbenveniste** [ivadomed/ms-lesion-agnostic#97]
I am investigating using DL methods for lesion tracking across timepoints. This investigation is particularly aimed at tackling the issues of lesions fusing dividing through time.

This figure from [this paper](https://pubmed.ncbi.nlm.nih.gov/38457838/) sums it up pretty nicely in another context than MS: 

<img width="724" height="409" alt="Image" src="https://github.com/user-attachments/assets/f1076914-586d-4baa-b3ff-f0fe2c9ce2b1" />

To deal with this and to avoid registration, I am using lesion coordinates in the anatomical space. Each lesion is described by: 
```python
'subject': subject # The subject of the lesion
'timepoint': timepoint1, # the timepoint of the scan: either M0 or M12
'group': lesion_info['group'], # The group to which the lesion belongs to (named tracks in the figure above)
'z': lesion_info['centerline_z'],  # The continuous vertebral-level of the center-of-mass (CoM) 
'r': lesion_info['radius_mm'], # The distance of CoM to the centerline
'theta': lesion_info['theta'], # The angle with the AP direction
'volume': lesion_info['volume_mm3'], # The volume of the lesion
'RL_diameter': lesion_info['RL_diameter_mm'], # The max diameter in the RL direction 
'AP_diameter': lesion_info['AP_diameter_mm'], # The max diameter in the AP direction
'IS_diameter': lesion_info['IS_diameter_mm'] # The max diameter in IS direction
```

To better understand the coordinates, here is a figure:

<img width="913" height="621" alt="Image" src="https://github.com/user-attachments/assets/3a29ac0d-cd8c-43ca-a9e7-fa5298c7a8ff" />

I build the dataset in the following way.
For each lesion at baseline of a subject, I create a pair with all the lesions from the follow-up. These lesion pairs are assigned a label 1 if they belong to the same group. I train ML models to solve this. 

Machine models investigated:
- XGBoost
- Siamese network
- GNN

**Comment by plbenveniste** [comment]
Since, I am training models, I had to split the dataset into training, validation and test. 
I arbitrarily decided on the following split:
- test: subjects from Toronto
- validation: subjects from Calgary and Montreal
- train: the rest

It adds up to the following:
```console
Total subjects: 34
Training subjects: 17, Validation subjects: 8, Testing subjects: 9
Training set size: 654 pairs, Validation set size: 314 pairs, Testing set size: 351 pairs
```

**Comment by plbenveniste** [comment]
## Siamese model training

I tried training a Siamese model to predict lesion groups.
The network takes as input two pairs of lesions (from the same subject but different timepoints) and computes their respective embeddings using a small MLP. It then predicts if they are from the same groups using a distance function.

### Specific parameters for this training

- Data are scaled (removing the mean and scaling to unit variance) using StandardScaler()
- The embeddings are computed using the following: 
```python
input_enc = Input(shape=(num_features_per_lesion,))
x = layers.Dense(128, activation='relu')(input_enc)
x = layers.BatchNormalization()(x)
x = layers.Dense(64, activation='relu')(x)
embedding = layers.Dense(32, activation='linear')(x)

encoder = models.Model(inputs=input_enc, outputs=embedding)
```
- an L2 distance is used: 
```python
L2_layer = layers.Lambda(lambda tensors: K.sqrt(K.sum(K.square(tensors[0] - tensors[1]), axis=1, keepdims=True)))
L2_distance = L2_layer([processed_a, processed_b])
prediction = layers.Dense(1, activation='sigmoid')(L2_distance)
```
- Optimizer:  `optimizer = optimizers.RMSprop(learning_rate=0.001)`
- Loss: `contrastive_loss_margin(margin=1)`
- Epoch: 500
- Batch size: 20
- weighting: weighting the batch of the training data to up-weight the positive class (to limit the unbalance of the dataset)
```console
Class weights: {0: np.float64(0.559931506849315), 1: np.float64(4.671428571428572)}
```


### Other configurations tried:

I tried other configurations for this training which did not provide better results:
- with or without data scaling
- bigger networks: (64,32,16), (31,16,8), (256,128,64)
- deeper networks: (256,128,64,32)
- L1 distance, L1 distance with a Dense layer, L2 with a Dense layer
- Binary Focal Crossentropy, Binary Crossentropy
- optimizer: Adam, AdamW
- lr: 1e-4, 1e-2
- Epoch: 1000
- Batch size: 10, 30 and 50

**Comment by plbenveniste** [comment]
## XGBoost model

In a later step, I also trained an XGBoost model taking lesion pairs (from same subject but different timepoints) as input, to predict whether the lesions belong to the same tracks. 

The features used were: 
```python
# For both lesions
'z': lesion_info['centerline_z'],  # The continuous vertebral-level of the center-of-mass (CoM) 
'r': lesion_info['radius_mm'], # The distance of CoM to the centerline
'theta': lesion_info['theta'], # The angle with the AP direction
'volume': lesion_info['volume_mm3'], # The volume of the lesion
'RL_diameter': lesion_info['RL_diameter_mm'], # The max diameter in the RL direction 
'AP_diameter': lesion_info['AP_diameter_mm'], # The max diameter in the AP direction
'IS_diameter': lesion_info['IS_diameter_mm'] # The max diameter in IS direction
# For the lesion pair
'dz': dz # The z difference between baseline and follow-up lesion
'dr': dr # The r difference between baseline and follow-up lesion
'dtheta': dtheta # The theta difference between baseline and follow-up lesion
'dvol': d_vol # The volume difference between baseline and follow-up lesion
'dist': dist # The distance between baseline and follow-up lesion
```

The XGBoost model was trained. with the following parameters:
- evaluation metric: logloss
- for optimization of hyperparameters, we used bayesian otptimisaion
  -  'max_depth': Integer(3, 10),
  - 'min_child_weight': Integer(1, 10),
  - 'subsample': Real(0.5, 1),
  - 'colsample_bytree': Real(0.001, 1),
  - 'learning_rate': Real(0.01, 0.5, prior='log-uniform'),
  - 'scale_pos_weight': Real(5, 10, prior='log-uniform')  # To handle class imbalance
  - n_iter=50
  - n_jobs=1
  - cv=3
  - scoring='average_precision'

The hyperparameters found were: 
```python
{'colsample_bytree': 0.8750152904321713,
                        'learning_rate': 0.0382641683481247,
                        'max_depth': 10,
                        'min_child_weight': 10,
                        'scale_pos_weight': 7.0202360215459345,
                        'subsample': 0.5}
```

**Comment by plbenveniste** [comment]
Models were trained using features extracted from manual segmentation of lesions.
Here are the results of the various methods using either the manual segmentation or the predicted lesions to extract the lesion features.

On manual lesion segmentation:
<img width="1503" height="451" alt="Image" src="https://github.com/user-attachments/assets/308f9686-8454-4976-935f-2633f547a98f" />

On predicted lesion segmentation:

<img width="1507" height="443" alt="Image" src="https://github.com/user-attachments/assets/fe8da9f5-cf2c-4a40-b636-7d1e8ec7f61b" />

Note:

- [x] I need to check why the Siamese train set on GT lesion has more TP+FN than the others.

**Comment by plbenveniste** [comment]
I have fixed the error for the incorrect counts : https://github.com/ivadomed/ms-lesion-agnostic/issues/98

**Comment by plbenveniste** [comment]
I have computed the performance of the various strategies while using only the predicted lesion segmentations for which GT mappings were manually created (https://github.com/ivadomed/ms-lesion-agnostic/issues/92#issuecomment-3627711141).

The performances are the following:

<img width="1805" height="576" alt="Image" src="https://github.com/user-attachments/assets/e9c8289d-44e4-4ec4-b9d4-d4490941e58f" />

**Comment by plbenveniste** [comment]
Closing this issue as investigations are finished. Code and investigation is saved in the release r20251208.
